src/Adapter/MSTest.TestAdapter/Discovery/AssemblyEnumerator.cs

```diff
@@ -92,17 +93,25 @@ internal ICollection<UnitTestElement> EnumerateAssembly(string assemblyFileName,
         var assembly = PlatformServiceProvider.Instance.FileOperations.LoadAssembly(assemblyFileName, isReflectionOnly: false);
 
         var types = GetTypes(assembly, assemblyFileName, warningMessages);
-        var discoverInternals = assembly.GetCustomAttribute<UTF.DiscoverInternalsAttribute>() != null;
-        var testDataSourceDiscovery = assembly.GetCustomAttribute<UTF.TestDataSourceDiscoveryAttribute>()?.DiscoveryOption ?? UTF.TestDataSourceDiscoveryOption.DuringDiscovery;
-
+        var discoverInternals = assembly.GetCustomAttribute<DiscoverInternalsAttribute>() != null;
+        var testIdGenerationStrategy = assembly.GetCustomAttribute<TestIdGenerationStrategyAttribute>()?.Strategy
+            ?? TestIdGenerationStrategy.FullyQualifiedTest;
+
+        var testDataSourceDiscovery = assembly.GetCustomAttribute<TestDataSourceDiscoveryAttribute>()?.DiscoveryOption
+#pragma warning disable CS0618 // Type or member is obsolete
+            ?? (testIdGenerationStrategy == TestIdGenerationStrategy.Legacy
```

Comment:
When using legacy strategy, there is no point in trying to "read" data during discovery as the ID will do nothing with them.